server/aws-lsp-codewhisperer/src/language-server/chat/chatEventParser.ts

```diff
@@ -0,0 +1,97 @@
+import { ChatResponseStream, Reference, SupplementaryWebLink } from '@amzn/codewhisperer-streaming'
+import {
+    ChatItemAction,
+    ChatResult,
+    ReferenceTrackerInformation,
+    SourceLink,
+} from '@aws/language-server-runtimes/protocol'
+
+export class ChatEventParser implements ChatResult {
+    error?: string
+    messageId?: string
+    body?: string
+    canBeVoted?: boolean
+    relatedContent?: { title?: string; content: SourceLink[] }
+    followUp?: { text?: string; options?: ChatItemAction[] }
+    codeReference?: ReferenceTrackerInformation[]
+
+    static getReferencedInformation(reference: Reference): string {
+        return `Reference code under **${reference.licenseName}** license from repository \`${reference.repository}\``
+    }
+
+    static mapRelatedData({ title = '', url = '', snippet }: SupplementaryWebLink): SourceLink {
+        return {
+            title,
+            url,
+            body: snippet,
+        }
+    }
+    static mapReferenceData(reference: Reference): ReferenceTrackerInformation {
+        return {
+            ...reference,
+            information: ChatEventParser.getReferencedInformation(reference),
+        }
+    }
+
+    constructor(messageId: string) {
+        this.messageId = messageId
+    }
+
+    public processPartialEvent(chatEvent: ChatResponseStream): ChatResult {
```

Comment:
You could convert `ChatResult` into `Result<T, TError>` too. That way, you would not need `error` field on the class. Usage of the class would be more intuitive.

Reply:
Agreed. I will update these to follow the same pattern in a follow up soon